File: app/schemas/team_event.py
import uuid
from datetime import datetime
from datetime import time as time_

from pydantic import BaseModel, Field

from app.models.team_event import (
    TeamEventAbsenceReason,
    TeamEventAttendanceStatus,
    TeamEventPublishStatus,
    TeamEventStatus,
    TeamEventType,
)
from app.models.user import Position


class TeamEventDrillRead(BaseModel):
    id: uuid.UUID
    section_id: uuid.UUID
    order: int
    title: str
    description: str | None = None
    duration_minutes: int | None = None


class TeamEventDrillSectionRead(BaseModel):
    id: uuid.UUID
    order: int
    name: str
    drills: list[TeamEventDrillRead]


class TeamEventRead(BaseModel):
    """Assembled manually in TeamEventService, not from_attributes --
    `sections` is None while the board is a draft and the caller isn't the
    captain (event exists, content hidden), [] once published with no
    sections yet, and populated once they exist. GAME events always get
    board_status=None/sections=None -- games have no board.
    """

    id: uuid.UUID
    team_id: uuid.UUID
    event_type: TeamEventType
    status: TeamEventStatus
    starts_at: datetime
    opponent_name: str | None = None
    board_status: TeamEventPublishStatus | None = None
    sections: list[TeamEventDrillSectionRead] | None = None
    created_at: datetime


class TeamEventCreate(BaseModel):
    event_type: TeamEventType
    starts_at: datetime
    opponent_name: str | None = Field(default=None, max_length=100)


class TeamEventReschedule(BaseModel):
    starts_at: datetime


class TeamEventDrillSectionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)


class TeamEventDrillSectionUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=100)


class TeamEventDrillSectionReorder(BaseModel):
    section_ids: list[uuid.UUID] = Field(min_length=1)


class TeamEventDrillCreate(BaseModel):
    section_id: uuid.UUID
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    duration_minutes: int | None = Field(default=None, ge=1, le=180)


class TeamEventDrillUpdate(BaseModel):
    # A different section_id moves the drill to the end of that section.
    section_id: uuid.UUID
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    duration_minutes: int | None = Field(default=None, ge=1, le=180)


class TeamEventDrillReorder(BaseModel):
    """Order within ONE section -- drill_ids must be exactly its drills."""

    section_id: uuid.UUID
    drill_ids: list[uuid.UUID] = Field(min_length=1)


class TeamEventAttendanceSet(BaseModel):
    status: TeamEventAttendanceStatus
    reason: TeamEventAbsenceReason | None = None
    reason_note: str | None = None


class TeamEventAttendanceRead(BaseModel):
    status: TeamEventAttendanceStatus
    reason: TeamEventAbsenceReason | None = None
    reason_note: str | None = None
    responded_at: datetime


class TeamEventAttendanceMemberRead(BaseModel):
    """One roster row -- assembled manually in TeamEventService from
    (TeamEventAttendance | None, User). reason/reason_note/responded_at
    stay None for the unmarked group.
    """

    user_id: uuid.UUID
    first_name: str
    last_name: str
    avatar_url: str | None = None
    reason: TeamEventAbsenceReason | None = None
    reason_note: str | None = None
    responded_at: datetime | None = None


class TeamEventAttendanceRosterRead(BaseModel):
    is_locked: bool
    going: list[TeamEventAttendanceMemberRead]
    not_going: list[TeamEventAttendanceMemberRead]
    unmarked: list[TeamEventAttendanceMemberRead]


class TeamEventNudgeResult(BaseModel):
    notified_count: int
    last_nudge_sent_at: datetime


class TeamEventLineupPlayerRead(BaseModel):
    """position is just a UI hint for forming game lines by role -- the
    backend never enforces it (see the v2 plan's section 7).
    """

    user_id: uuid.UUID
    first_name: str
    last_name: str
    avatar_url: str | None = None
    position: Position | None = None


class TeamEventLineupGroupRead(BaseModel):
    id: uuid.UUID
    name: str | None = None
    color: str | None = None
    players: list[TeamEventLineupPlayerRead]


class TeamEventLineupRead(BaseModel):
    """Assembled manually, same visibility contract as TeamEventRead.sections
    -- groups/unassigned are None while the lineup is a draft and the
    caller isn't the captain, populated once published (or always for the
    captain).
    """

    lineup_status: TeamEventPublishStatus
    groups: list[TeamEventLineupGroupRead] | None = None
    unassigned: list[TeamEventLineupPlayerRead] | None = None


class TeamEventLineupGroupCreate(BaseModel):
    name: str | None = Field(default=None, max_length=100)
    color: str | None = Field(default=None, max_length=20)


class TeamEventLineupGroupUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=100)
    color: str | None = Field(default=None, max_length=20)


class TeamEventLineupPlayerAssign(BaseModel):
    group_id: uuid.UUID


class TeamEventDiaryEntrySave(BaseModel):
    note: str | None = None


class TeamEventDiaryEntryRead(BaseModel):
    note: str | None = None
    created_at: datetime
    updated_at: datetime


class TeamIceScheduleTemplateRead(BaseModel):
    id: uuid.UUID
    weekday: int
    start_time: time_
    active: bool


class TeamIceScheduleTemplateCreate(BaseModel):
    weekday: int = Field(ge=0, le=6)
    start_time: time_


class TeamIceScheduleTemplateUpdate(BaseModel):
    active: bool

File: frontend/src/types/teamEvent.ts
import type { Position } from './user'

export const TEAM_EVENT_TYPES = ['training', 'game'] as const
export type TeamEventType = (typeof TEAM_EVENT_TYPES)[number]

export const TEAM_EVENT_STATUSES = ['scheduled', 'cancelled'] as const
export type TeamEventStatus = (typeof TEAM_EVENT_STATUSES)[number]

export const TEAM_EVENT_PUBLISH_STATUSES = ['draft', 'published'] as const
export type TeamEventPublishStatus = (typeof TEAM_EVENT_PUBLISH_STATUSES)[number]

export const TEAM_EVENT_ATTENDANCE_STATUSES = ['going', 'not_going'] as const
export type TeamEventAttendanceStatus = (typeof TEAM_EVENT_ATTENDANCE_STATUSES)[number]

export const TEAM_EVENT_ABSENCE_REASONS = ['work', 'injury', 'study', 'other'] as const
export type TeamEventAbsenceReason = (typeof TEAM_EVENT_ABSENCE_REASONS)[number]

export interface TeamEventDrillRead {
  id: string
  section_id: string
  order: number
  title: string
  description: string | null
  duration_minutes: number | null
}

// A named block of the board -- the coach names it freely; these are just
// the one-tap suggestions the editor offers.
export interface TeamEventDrillSectionRead {
  id: string
  order: number
  name: string
  drills: TeamEventDrillRead[]
}

export const DRILL_SECTION_PRESETS = ['Разминка', 'Катание', 'Броски', 'Игровые', 'Заминка'] as const

// GET/POST /teams/{team_id}/events -- sections is null while the board is a
// draft and the caller isn't the captain (event exists, content hidden),
// [] once published with nothing on it yet. GAME events always have
// board_status=null/sections=null -- games have no board.
export interface TeamEventRead {
  id: string
  team_id: string
  event_type: TeamEventType
  status: TeamEventStatus
  starts_at: string
  opponent_name: string | null
  board_status: TeamEventPublishStatus | null
  sections: TeamEventDrillSectionRead[] | null
  created_at: string
}

export interface TeamEventCreatePayload {
  event_type: TeamEventType
  starts_at: string
  opponent_name?: string | null
}

export interface TeamEventReschedulePayload {
  starts_at: string
}

export interface TeamEventDrillCreatePayload {
  section_id: string
  title: string
  description?: string | null
  duration_minutes?: number | null
}

// A different section_id moves the drill to the end of that section.
export interface TeamEventDrillUpdatePayload {
  section_id: string
  title: string
  description?: string | null
  duration_minutes?: number | null
}

export interface TeamEventAttendanceSetPayload {
  status: TeamEventAttendanceStatus
  reason?: TeamEventAbsenceReason | null
  reason_note?: string | null
}

export interface TeamEventAttendanceRead {
  status: TeamEventAttendanceStatus
  reason: TeamEventAbsenceReason | null
  reason_note: string | null
  responded_at: string
}

export interface TeamEventAttendanceMemberRead {
  user_id: string
  first_name: string
  last_name: string
  avatar_url: string | null
  reason: TeamEventAbsenceReason | null
  reason_note: string | null
  responded_at: string | null
}

export interface TeamEventAttendanceRosterRead {
  is_locked: boolean
  going: TeamEventAttendanceMemberRead[]
  not_going: TeamEventAttendanceMemberRead[]
  unmarked: TeamEventAttendanceMemberRead[]
}

export interface TeamEventNudgeResult {
  notified_count: number
  last_nudge_sent_at: string
}

export interface TeamEventLineupPlayerRead {
  user_id: string
  first_name: string
  last_name: string
  avatar_url: string | null
  position: Position | null
}

export interface TeamEventLineupGroupRead {
  id: string
  name: string | null
  color: string | null
  players: TeamEventLineupPlayerRead[]
}

// groups/unassigned are null while the lineup is a draft and the caller
// isn't the captain -- same visibility contract as TeamEventRead.drills.
export interface TeamEventLineupRead {
  lineup_status: TeamEventPublishStatus
  groups: TeamEventLineupGroupRead[] | null
  unassigned: TeamEventLineupPlayerRead[] | null
}

export interface TeamEventLineupGroupCreatePayload {
  name?: string | null
  color?: string | null
}

export interface TeamEventLineupGroupUpdatePayload {
  name?: string | null
  color?: string | null
}

export interface TeamEventDiaryEntryRead {
  note: string | null
  created_at: string
  updated_at: string
}

export interface TeamEventDiaryEntrySavePayload {
  note?: string | null
}

export interface TeamIceScheduleTemplateRead {
  id: string
  weekday: number
  start_time: string
  active: boolean
}

export interface TeamIceScheduleTemplateCreatePayload {
  weekday: number
  start_time: string
}
